Dockerfile

```diff
@@ -16,28 +21,37 @@ ENV PYTHONUNBUFFERED=1 \
     POETRY_NO_INTERACTION=1 \
     VENV_PATH="/.venv"
 
-ENV PATH="$VENV_PATH/bin:$PATH"
+# ugly hack for C extension from lru-dict
+ENV CFLAGS="-g0 -O2 -ffile-prefix-map=/src=."
 
-FROM base as builder
+ENV PATH="$VENV_PATH/bin:$PATH"
 
 ENV POETRY_VERSION=1.3.2
-RUN pip install --no-cache-dir poetry==$POETRY_VERSION
+
+RUN pip install --no-compile --no-cache-dir poetry==$POETRY_VERSION
 
 WORKDIR /
 COPY pyproject.toml poetry.lock ./
-RUN poetry install --only main --no-root
+RUN poetry install --only main --no-root --no-cache
 
+RUN rm -rf /root/.cache $VENV_PATH/src \
+  && find $VENV_PATH -type f -name 'RECORD' -delete \
+  && find $VENV_PATH -type f -name '*.pyc' -delete \
+  && find /root/.local -type f -name '*.pyc' -delete \
+  && find /usr/local -type f -name '*.pyc' -delete
 
 FROM base as production
 
+ENV VENV_PATH="/.venv"
+
 COPY --from=builder $VENV_PATH $VENV_PATH
 WORKDIR /app
 COPY . .
 
 RUN apt-get clean && find /var/lib/apt/lists/ -type f -delete && chown -R www-data /app/
```

Comment:
And this should be removed in the 'RUN' with apt-install.

Here the sample of one of CI files I do. I usually do it not for reproducible reasons, but for smaller image (so I don't want additional files in any layer, which is, the same goal as with reproducible images).

https://github.com/lidofinance/ci-image/blob/master/Dockerfile

Note how big is RUN stanza for apt and poetry.

In this particular case we also want to purge away python3.11-dev, so both apt and poetry are in the same RUN (layer).